### PR TITLE
DShot telemetry support (WIIIIIP)

### PR DESCRIPTION
**Pull request by glowtape**
Early version, posted for appraisal while I'm trying to figure out something unrelated, and to see whether we want or need this (UAVO is 150 bytes). Seems to work with a single ESC. Need to do some more work on handling timeouts.

Surprisingly there's at least one thing not up to spec with BLHeli32 (current isn't output in the units the KISS specs wants).

### CODE REVIEW

**Comment by glowtape** [comment]
Hooked up telemetry wires to the other ESCs, since I had to switch out motors anyway. Seems to capture telemetry for the appropriate actuators properly.

**Comment by tracernz** [comment]
Neat. I'd been wondering about doing this. 👍

**Comment by glowtape** [comment]
Voltage and mAh Consumed could probably turned into a single field. Since all ESCs usually draw from the same power bus, it doesn't make sense to keep separate voltages. Similar for consumed energy. Would also shrink the UAVO size.

While trying to make a screenshot for pudding, I noticed that it keeps timing out (currently 7.5ms in my local branch) when the motors are spinning, but doesn't appear to when they're disarmed. Not sure where to move the blame, need to put the scope on it.

Also, once the timeout business is solved, and I got some clarification from the BLHeli guys regarding the current values, might be an idea to let the Battery Monitoring module tap alternatively into the telemetry UAVO, if there aren't any sensors.

Either way, here's the aforementioned pudding:

esctelem

**Comment by glowtape** [comment]
Collapsed voltage and mAh consumed, UAVO dropped from 150 to 96 bytes. Polls every 200ms now. Some history rewriting.

File: flight/PiOS/STM32/pios_servo.c [hunk 1]
```diff
@@ -99,6 +99,8 @@ static bool resetting;
 
 static bool dshot_in_use;
 
+static int telemreq_servo;
```
Comment (mlyle) [inline flight/PiOS/STM32/pios_servo.c]: At the least this needs to be volatile, but it probably needs to be atomic in actuality.  This also deepens the debt for flyingpi.

**Comment by glowtape** [comment]
Jenkins, build artifacts please.

**Comment by glowtape** [comment]
@ufoDziner Seems to work on the desk. Since this doesn't contain the LQG stuff, it'll trash some of the configuration. Best to save the settings partition of the LQG configured quads before switching to this.

**Comment by ufoDziner** [comment]
Appears to work with X-Racer Quadrant 25A escs.  Will continue to test.

**Comment by tracernz** [comment]
KISS is the reference implementation right? Should be able to rig that up later in the week.

**Comment by glowtape** [comment]
Yeah, KISS. BLHeli implements it the same (for now) from what I can tell.

**Comment by ufoDziner** [comment]
@glowtape Does this require any further testing for blheli_32?

**Comment by glowtape** [comment]
Not that I know of. I had it running on _32, you on _s, so it works. I need to eventually prettify it a bit.

File: flight/Modules/Actuator/actuator.c [hunk 2]
```diff
@@ -251,6 +282,12 @@ int32_t ActuatorInitialize()
 	}
 #endif
 
+#if defined(PIOS_INCLUDE_ESCTELEMETRY)
+	if (ActuatorTelemetryInitialize() == -1) {
```
Comment (tracernz) [inline flight/Modules/Actuator/actuator.c]: Safe to go `< 0` or `!= 0` IMO, *Initialize might return multiple error codes at some point. :wink: Actually, [it already can return `-2`](https://github.com/d-ronin/dRonin/blob/next/flight/UAVObjects/uavobjecttemplate.c#L57), so this code is wrong.

File: flight/Modules/Actuator/actuator.c [hunk 3]
```diff
@@ -611,6 +648,80 @@ static void post_process_scale_and_commit(float *motor_vect,
...
 
+#if defined(PIOS_INCLUDE_ESCTELEMETRY)
+	if (PIOS_ESCTelemetry_IsAvailable()) {
+
+		/* Process telemetry request stuff before servo settings. */
+
+		if (!tlm) {
+			tlm = PIOS_malloc_no_dma(sizeof(*tlm));
+			if (!tlm)
+				PIOS_Assert(0);
```
Comment (tracernz) [inline flight/Modules/Actuator/actuator.c]: PIOS_Assert(tlm);?

File: flight/Modules/Actuator/actuator.c [hunk 4]
```diff
@@ -611,6 +648,80 @@ static void post_process_scale_and_commit(float *motor_vect,
...
+			/* If active servo is 0xFF, no request is active, either because we're starting over,
+			   the servo doesn't exists, or isn't configured to DShot. */
+
+			tlm->active = tlm->servo;
+			tlm->servo++;
+
+			tlm->req_start = PIOS_DELAY_GetuS();
+
+			/* We don't care about ongoing requests, because we shouldn't even be here, if one is. */
+			if (PIOS_Servo_RequestTelemetry(tlm->active) < -1)
```
Comment (tracernz) [inline flight/Modules/Actuator/actuator.c]: What does -1 mean? I always prefer to use enums when actually differentiating errors. e.g.
`if (PIOS_*() == -EBUSY)` or similar is a lot more readable than `if (PIOS_*() == -5)`.

File: flight/PiOS/Common/pios_esctelemetry.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+				// Unfreeze telemetry driver.
+				break;
+			}
+		}
+	}
+
+	dev->rx_timer = 0;
+
+	// Keep pumping data.
+	if(headroom) *headroom = ESC_TELEMETRY_LENGTH;
```
Comment (tracernz) [inline flight/PiOS/Common/pios_esctelemetry.c]: New lines, and space after if (applies to next few lines too).

File: flight/PiOS/Common/pios_esctelemetry.c [hunk 6]
```diff
@@ -0,0 +1,244 @@
+/**
```
Comment (tracernz) [inline flight/PiOS/Common/pios_esctelemetry.c]: This design will not work for multiple different ESC protocols? This should be `PIOS_DShot_Telemetry` or something? And other ESC telemetry drivers need a way to hook in somehow. Other protocols existed before DShot, and we have already discussed supporting them, so a generic `PIOS_ESCTelemetry` implementation that can hook into multiple different backend drivers is essential IMO.

File: flight/PiOS/Common/pios_esctelemetry.c [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+	PIOS_ESCTelemetry_ResetBuffer(dev);
+
+	uint8_t crc = PIOS_CRC_updateCRC(0, dev->rx_buf, ESC_TELEMETRY_LENGTH-1);
+
+	if (crc == dev->rx_buf[ESC_TELEMETRY_LENGTH-1]) {
+
+		dev->telem.temperature = dev->rx_buf[0];
+		dev->telem.voltage = (float)(dev->rx_buf[1]<<8 | dev->rx_buf[2]) * 0.01f;
+		dev->telem.current = (float)(dev->rx_buf[3]<<8 | dev->rx_buf[4]) * 0.01f;
+		dev->telem.mAh = (uint16_t)(dev->rx_buf[5]<<8 | dev->rx_buf[6]);
```
Comment (tracernz) [inline flight/PiOS/Common/pios_esctelemetry.c]: Missing some spaces around operators on these lines.

**Review by tracernz — CHANGES_REQUESTED** [review]
Nice work overall. :+1: Apologies for taking so long to review.

**Comment by ufoDziner** [comment]
@glowtape Actually mine was _32 as well. I don't think _s supports it.

File: flight/Modules/Actuator/actuator.c (hunk 2 above)
Comment (glowtape) [reply flight/Modules/Actuator/actuator.c]: Value -2 means already initialized, which isn't a reason we'd like to fail for. The all other UAVO inits do the == -1 thing, too.

File: flight/PiOS/Common/pios_esctelemetry.c (hunk 6 above)
Comment (glowtape) [reply flight/PiOS/Common/pios_esctelemetry.c]: This round gonna do it just DShot specific and rename the current driver accordingly. Later on we could revisit this and split things into generic parts and actuator specific telemetry. Haven't thought it through where to draw the line yet.

**Comment by glowtape** [comment]
Rebased, applied review stuff. Added actuator "pins" to the Battery module to get data from the ActuatorTelemetry UAVO instead.

**Comment by ufoDziner** [comment]
Jenkins, build artifacts please.

**Comment by jihlein** [comment]
Were does this effort stand, does it need more testing?

**Comment by glowtape** [comment]
Everything's moving to Bi-DShot, or whatever the hell it is called, which sends telemetry back over the PWM wire instead of a separate serial wire. Currently slowly dabbling with making it work, so this is obsolete.

**Pull request closed without merging**